invenio_users_resources/records/models.py

```diff
@@ -25,103 +27,213 @@ class MockModel(dict, ABC):
     that's expected by API classes as far as necessary.
     """
 
-    def __init__(self, data=None, model_obj=None, **kwargs):
+    _properties = []
+    """Properties of this object that can be accessed."""
+
+    _set_properties = []
+    """Properties of this object that can be set."""
+
+    _data = None
+
+    def __init__(self, model_obj=None, **kwargs):
         """Constructor."""
-        data.update(kwargs)
-        super().__init__(data)
-        self._model_obj = model_obj
+        super().__setattr__("_data", {})
+        if model_obj is not None:
+            # E.g. when data is loaded from database
+            self.from_model(model_obj)
+            super().__setattr__("_model_obj", model_obj)
+        else:
+            # E.g. when data is loaded from the search index
+            self.from_kwargs(kwargs)
+            super().__setattr__("_model_obj", None)
 
     @property
     @abstractmethod
     def model_obj(self):
         """The actual model object behind this mock model."""
         return None
 
-    @property
-    def id(self):
-        """User identifier."""
-        return self.data["id"]
-
-    @property
-    def created(self):
-        """When the user was created."""
-        return self.model_obj.created
-
-    @property
-    def updated(self):
-        """When the user was last updated."""
-        return self.model_obj.updated
-
-    @property
-    def version_id(self):
-        """Used by SQLAlchemy for optimistic concurrency control."""
-        return self.model_obj.version_id
-
-    @property
-    def data(self):
-        """Get the user's data by decoding the JSON."""
-        return dict(self)
-
-    @property
-    def json(self):
-        """Provide the user's data as a JSON/dict blob."""
-        return dict(self)
-
+    def from_kwargs(self, kwargs):
+        """Extract information from kwargs."""
+        for p in self._properties:
+            self._data[p] = kwargs.get(p, None)
```

Comment:
```suggestion
self._data = {p: kwargs.get(p, None) for p in self._properties}
```

Reply:
I had to revert back to original code because we're overwriting the ``__setattr__`` that intercepts the assignment ``self._data = ...`` but not ``self._data[p]``. We could use ``super().__setattr__('_data', {...})`` instead but I don't think that's reads very well.